src/onegov/org/models/search.py

```diff
@@ -159,5 +166,221 @@ def subset_count(self) -> int:
+        self.web_search = query
+        self.page = page  # page index
+
+        self.nbr_of_docs = 0
+        self.nbr_of_results = 0
+
+    @cached_property
+    def available_documents(self) -> int:
+        if not self.nbr_of_docs:
+            _ = self.load_batch_results
+        return self.nbr_of_docs
+
+    @cached_property
+    def available_results(self) -> int:
+        if not self.nbr_of_results:
+            _ = self.load_batch_results
+        return self.nbr_of_results
+
+    @property
+    def q(self) -> str:
+        """
+        Returns the user's query term from the search field of the UI
+
+        """
+        return self.web_search
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, SearchPostgres):
+            return NotImplemented
+        return self.page == other.page and self.web_search == other.web_search
+
+    def subset(self) -> 'list[Searchable] | None':  # type:ignore[override]
+        return self.batch
+
+    @property
+    def page_index(self) -> int:
+        return self.page
+
+    def page_by_index(self, index: int) -> 'SearchPostgres[_M]':
+        return SearchPostgres(self.request, self.web_search, index)
+
+    @cached_property
+    def batch(self) -> 'list[Searchable]':  # type:ignore[override]
+        if not self.web_search:
+            return []
+
+        if self.web_search.startswith('#'):
+            results = self.hashtag_search()
+        else:
+            results = self.generic_search()
+
+        return results[self.offset:self.offset + self.batch_size]
+
+    @cached_property
+    def load_batch_results(self) -> list[Any]:
+        """
+        Load search results and sort upcoming events by occurrence start date.
+        This methods is a wrapper around `batch.load()`, which returns the
+        actual search results form the query.
+
+        """
+        batch: list[Searchable] = self.batch
+        future_events: list[Searchable] = []
+        other: list[Searchable] = []
+
+        for search_result in batch:
+            if (isinstance(search_result, Event)
+                    and search_result.latest_occurrence
+                    and search_result.latest_occurrence.start > utcnow()):
+                future_events.append(search_result)
+            else:
+                other.append(search_result)
+
+        if not future_events:
+            return batch
+
+        sorted_events = sorted(
+            future_events, key=lambda e:
+            e.latest_occurrence.start,  # type:ignore[attr-defined]
+            reverse=True)
+
+        return sorted_events + other
+
+    def _create_weighted_vector(
+        self,
+        model: Any,
+        language: str = 'simple'
+    ) -> Any:
+        # for now weight the first field with 'A', the rest with 'B'
+        weighted_vectors = [
+            func.setweight(
+                func.to_tsvector(
+                    language,
+                    getattr(model.fts_idx_data, field, '')),
```

Comment:
the weighted vector bases on the static data from column `fts_idx_data` generated upon `update` or `reindex` events.